src/java.net.http/share/classes/jdk/internal/net/http/PullPublisher.java

```diff
@@ -25,41 +25,45 @@
 
 package jdk.internal.net.http;
 
-import java.util.Iterator;
 import java.util.concurrent.Flow;
+
 import jdk.internal.net.http.common.Demand;
 import jdk.internal.net.http.common.SequentialScheduler;
 
 /**
- * A Publisher that publishes items obtained from the given Iterable. Each new
- * subscription gets a new Iterator.
+ * A {@linkplain Flow.Publisher publisher} that publishes items obtained from the given {@link CheckedIterator} supplier.
+ * Each new subscription gets a new {@code CheckedIterator}.
  */
 class PullPublisher<T> implements Flow.Publisher<T> {
 
     // Only one of `iterable` and `throwable` can be non-null. throwable is
     // non-null when an error has been encountered, by the creator of
     // PullPublisher, while subscribing the subscriber, but before subscribe has
     // completed.
-    private final Iterable<T> iterable;
+    private final CheckedIterable<T> iterable;
     private final Throwable throwable;
 
-    PullPublisher(Iterable<T> iterable, Throwable throwable) {
+    PullPublisher(CheckedIterable<T> iterable, Throwable throwable) {
+        if ((iterable == null && throwable == null) || (iterable != null && throwable != null)) {
```

Comment:
You can express this as `((iterable == null) == (throwable == null))` for simplicity.